File: local-host/src/agent-spike/rpc.ts
/**
 * LSP-style JSON-RPC 2.0 over a Node socket.
 *
 * Bidirectional. Both sides can `call` (outbound request) and `notify`
 * (outbound notification) and register handlers for incoming requests.
 * Wire format: `Content-Length: N\r\n\r\n{N bytes of UTF-8 JSON}`.
 */

import { randomUUID } from "node:crypto";
import { Socket } from "node:net";

export type RpcHandler = (params: any) => Promise<any> | any;

export class RpcError extends Error {
  constructor(public code: number, message: string, public data?: unknown) {
    super(message);
    this.name = "RpcError";
  }
}

interface PendingCall {
  resolve: (value: any) => void;
  reject: (error: unknown) => void;
  timer?: NodeJS.Timeout;
}

export class RpcEndpoint {
  private buf: Buffer<ArrayBufferLike> = Buffer.alloc(0);
  private handlers = new Map<string, RpcHandler>();
  private pending = new Map<string, PendingCall>();
  private closed = false;
  private onNotificationFn: ((method: string, params: any) => void) | null = null;
  private closeListeners: Array<() => void> = [];

  constructor(private socket: Socket) {
    socket.on("data", (chunk) => this.onData(chunk));
    socket.on("close", () => this.onClose());
    socket.on("error", (err) => {
      this.onClose();
      // surface errors via close; nothing else to do here
      void err;
    });
  }

  register(method: string, handler: RpcHandler): void {
    this.handlers.set(method, handler);
  }

  onNotification(fn: (method: string, params: any) => void): void {
    this.onNotificationFn = fn;
  }

  onClose(): void {
    if (this.closed) return;
    this.closed = true;
    for (const [, p] of this.pending) {
      if (p.timer) clearTimeout(p.timer);
      p.reject(new RpcError(-32000, "peer closed"));
    }
    this.pending.clear();
    for (const fn of this.closeListeners) {
      try {
        fn();
      } catch {
        // ignore — one listener should not break the others
      }
    }
    this.closeListeners = [];
  }

  whenClosed(fn: () => void): void {
    if (this.closed) {
      fn();
      return;
    }
    this.closeListeners.push(fn);
  }

  get isClosed(): boolean {
    return this.closed;
  }

  async call(method: string, params: unknown, timeoutMs = 30_000): Promise<any> {
    if (this.closed) throw new RpcError(-32000, "endpoint closed");
    const id = randomUUID();
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new RpcError(-32001, `call ${method} timed out after ${timeoutMs}ms`));
      }, timeoutMs);
      this.pending.set(id, { resolve, reject, timer });
      this.send({ jsonrpc: "2.0", id, method, params });
    });
  }

  notify(method: string, params: unknown): void {
    if (this.closed) return;
    this.send({ jsonrpc: "2.0", method, params });
  }

  close(): void {
    if (!this.closed) {
      this.socket.end();
    }
  }

  // --- private ---

  private send(msg: object): void {
    const body = Buffer.from(JSON.stringify(msg), "utf8");
    const header = Buffer.from(`Content-Length: ${body.length}\r\n\r\n`, "ascii");
    this.socket.write(Buffer.concat([header, body]));
  }

  private onData(chunk: Buffer): void {
    this.buf = this.buf.length === 0 ? chunk : Buffer.concat([this.buf, chunk]);
    while (true) {
      const headerEnd = this.buf.indexOf("\r\n\r\n");
      if (headerEnd === -1) return;
      const headerStr = this.buf.subarray(0, headerEnd).toString("ascii");
      let length = -1;
      for (const line of headerStr.split("\r\n")) {
        const ix = line.indexOf(":");
        if (ix === -1) continue;
        if (line.slice(0, ix).trim().toLowerCase() === "content-length") {
          length = Number(line.slice(ix + 1).trim());
        }
      }
      if (length < 0 || !Number.isFinite(length)) {
        // malformed; drop everything up through header to try to recover
        this.buf = this.buf.subarray(headerEnd + 4);
        continue;
      }
      const bodyStart = headerEnd + 4;
      if (this.buf.length < bodyStart + length) return;
      const body = this.buf.subarray(bodyStart, bodyStart + length);
      this.buf = this.buf.subarray(bodyStart + length);
      let msg: any;
      try {
        msg = JSON.parse(body.toString("utf8"));
      } catch {
        continue;
      }
      this.dispatch(msg);
    }
  }

  private dispatch(msg: any): void {
    // response to one of our outbound calls?
    if (
      typeof msg.id === "string" &&
      typeof msg.method === "undefined" &&
      ("result" in msg || "error" in msg)
    ) {
      const p = this.pending.get(msg.id);
      if (!p) return;
      this.pending.delete(msg.id);
      if (p.timer) clearTimeout(p.timer);
      if (msg.error) {
        p.reject(new RpcError(msg.error.code ?? -32000, msg.error.message ?? "", msg.error.data));
      } else {
        p.resolve(msg.result);
      }
      return;
    }

    const method = msg.method as string | undefined;
    const params = msg.params ?? {};
    const id = msg.id as string | undefined;

    if (!method) return;
    const handler = this.handlers.get(method);

    if (!handler) {
      if (id !== undefined) {
        this.send({
          jsonrpc: "2.0",
          id,
          error: { code: -32601, message: `method not found: ${method}` },
        });
      } else {
        this.onNotificationFn?.(method, params);
      }
      return;
    }

    void Promise.resolve()
      .then(() => handler(params))
      .then(
        (result) => {
          if (id !== undefined) {
            this.send({ jsonrpc: "2.0", id, result: result ?? null });
          }
        },
        (err: unknown) => {
          if (id === undefined) return;
          if (err instanceof RpcError) {
            this.send({
              jsonrpc: "2.0",
              id,
              error: { code: err.code, message: err.message, data: err.data },
            });
          } else {
            const message = err instanceof Error ? err.message : String(err);
            this.send({ jsonrpc: "2.0", id, error: { code: -32000, message } });
          }
        }
      );
  }
}

File: local-host/python-spike/rpc.py
"""LSP-style JSON-RPC 2.0 over Unix Domain Socket.

Bidirectional: a single endpoint can both serve incoming requests and
make outbound calls. Notifications (no id) get no reply. Each frame is:

    Content-Length: N\r\n
    \r\n
    {N bytes of JSON}
"""

from __future__ import annotations

import asyncio
import json
import logging
from collections.abc import Awaitable, Callable
from typing import Any
from uuid import uuid4

log = logging.getLogger("rpc")

Handler = Callable[[dict[str, Any]], Awaitable[Any]]


class RpcError(Exception):
    def __init__(self, code: int, message: str, data: Any = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.data = data


class RpcEndpoint:
    def __init__(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
    ) -> None:
        self._reader = reader
        self._writer = writer
        self._handlers: dict[str, Handler] = {}
        self._pending: dict[str, asyncio.Future[Any]] = {}
        self._write_lock = asyncio.Lock()
        self._closed = asyncio.Event()

    # --------- public API ---------

    def register(self, method: str, handler: Handler) -> None:
        self._handlers[method] = handler

    async def call(self, method: str, params: Any = None, timeout: float = 30.0) -> Any:
        req_id = uuid4().hex
        fut: asyncio.Future[Any] = asyncio.get_running_loop().create_future()
        self._pending[req_id] = fut
        await self._send({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})
        try:
            return await asyncio.wait_for(fut, timeout)
        finally:
            self._pending.pop(req_id, None)

    async def notify(self, method: str, params: Any = None) -> None:
        await self._send({"jsonrpc": "2.0", "method": method, "params": params})

    async def run(self) -> None:
        try:
            while not self._closed.is_set():
                msg = await self._read_message()
                if msg is None:
                    break
                asyncio.create_task(self._dispatch(msg))
        finally:
            self.close()

    def close(self) -> None:
        self._closed.set()
        try:
            self._writer.close()
        except Exception:
            pass
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(RpcError(-32000, "peer closed"))
        self._pending.clear()

    @property
    def closed(self) -> bool:
        return self._closed.is_set()

    # --------- internals ---------

    async def _send(self, msg: dict[str, Any]) -> None:
        body = json.dumps(msg, separators=(",", ":")).encode("utf-8")
        header = f"Content-Length: {len(body)}\r\n\r\n".encode("ascii")
        async with self._write_lock:
            self._writer.write(header + body)
            await self._writer.drain()

    async def _read_message(self) -> dict[str, Any] | None:
        headers: dict[str, str] = {}
        while True:
            line = await self._reader.readline()
            if not line:
                return None
            line = line.rstrip(b"\r\n")
            if not line:
                break
            try:
                key, value = line.decode("ascii").split(":", 1)
            except ValueError:
                continue
            headers[key.strip().lower()] = value.strip()
        length_s = headers.get("content-length")
        if length_s is None:
            return None
        length = int(length_s)
        body = await self._reader.readexactly(length)
        return json.loads(body.decode("utf-8"))

    async def _dispatch(self, msg: dict[str, Any]) -> None:
        # response to one of our outbound calls?
        if "id" in msg and ("result" in msg or "error" in msg) and "method" not in msg:
            fut = self._pending.get(msg["id"])
            if fut is None or fut.done():
                return
            if "error" in msg:
                err = msg["error"]
                fut.set_exception(RpcError(err.get("code", -32000), err.get("message", ""), err.get("data")))
            else:
                fut.set_result(msg.get("result"))
            return

        # incoming request or notification
        method = msg.get("method")
        params = msg.get("params")
        req_id = msg.get("id")
        handler = self._handlers.get(method or "")

        if handler is None:
            if req_id is not None:
                await self._send(
                    {
                        "jsonrpc": "2.0",
                        "id": req_id,
                        "error": {"code": -32601, "message": f"method not found: {method}"},
                    }
                )
            else:
                log.warning("unhandled notification: %s", method)
            return

        try:
            result = await handler(params or {})
        except RpcError as e:
            if req_id is not None:
                await self._send(
                    {
                        "jsonrpc": "2.0",
                        "id": req_id,
                        "error": {"code": e.code, "message": e.message, "data": e.data},
                    }
                )
            return
        except Exception as e:
            log.exception("handler %s raised", method)
            if req_id is not None:
                await self._send(
                    {
                        "jsonrpc": "2.0",
                        "id": req_id,
                        "error": {"code": -32000, "message": str(e)},
                    }
                )
            return

        if req_id is not None:
            await self._send({"jsonrpc": "2.0", "id": req_id, "result": result})
